// @claude-flow/neural/dist/utils/serialize.js
/**
 * JSON-safe serialization helpers for neural state.
 *
 * The package's hot-path types use `Float32Array` and `Map<string, X>` which
 * don't survive `JSON.stringify` cleanly. These helpers encode/decode them
 * in a stable shape so SONAManager / ReasoningBank / PatternLearner can
 * round-trip through `serialize()` → `JSON.stringify` → file → `JSON.parse`
 * → `deserialize()` without losing precision or structure.
 *
 * Encoding:
 *   Float32Array  ↔ { __f32: number[] }
 *   Map<K, V>     ↔ { __map: Array<[K, V]> }
 *
 * Numbers in Float32Array are stored as plain JS `number` (double precision).
 * Round-trip is loss-free at single-precision (the values were f32 originally;
 * widening to f64 then narrowing back is exact for any finite f32).
 */
export function encodeFloat32Array(arr) {
    return { __f32: Array.from(arr) };
}
export function decodeFloat32Array(encoded) {
    return new Float32Array(encoded.__f32);
}
export function isEncodedFloat32Array(v) {
    return typeof v === 'object' && v !== null && Array.isArray(v.__f32);
}
export function encodeMap(map, encodeValue) {
    const entries = [];
    for (const [k, v] of map.entries()) {
        entries.push([k, encodeValue ? encodeValue(v) : v]);
    }
    return { __map: entries };
}
export function decodeMap(encoded, decodeValue) {
    const m = new Map();
    for (const [k, v] of encoded.__map) {
        m.set(k, (decodeValue ? decodeValue(v) : v));
    }
    return m;
}
export function isEncodedMap(v) {
    return typeof v === 'object' && v !== null && Array.isArray(v.__map);
}
/**
 * Deep walk an object and encode all Float32Array/Map nodes. Useful when
 * a class's state contains nested f32 arrays in unknown locations (e.g.,
 * Trajectory.steps[i].stateEmbedding). Skip with caution — the deep walker
 * is recursive and shouldn't be used on objects with cycles.
 */
export function deepEncode(value) {
    if (value instanceof Float32Array)
        return encodeFloat32Array(value);
    if (value instanceof Map) {
        return encodeMap(value, (v) => deepEncode(v));
    }
    if (Array.isArray(value))
        return value.map((v) => deepEncode(v));
    if (value !== null && typeof value === 'object') {
        const out = {};
        for (const [k, v] of Object.entries(value))
            out[k] = deepEncode(v);
        return out;
    }
    return value;
}
export function deepDecode(value) {
    if (isEncodedFloat32Array(value))
        return decodeFloat32Array(value);
    if (isEncodedMap(value)) {
        return decodeMap(value, (v) => deepDecode(v));
    }
    if (Array.isArray(value))
        return value.map((v) => deepDecode(v));
    if (value !== null && typeof value === 'object') {
        const out = {};
        for (const [k, v] of Object.entries(value))
            out[k] = deepDecode(v);
        return out;
    }
    return value;
}
